File: InMoto/Models/MotoRoute.swift
import Foundation

/// Mezzo con cui si percorre il tragitto: determina il tipo di routing e la
/// taratura dei tempi di percorrenza.
enum TransportMode: String, CaseIterable, Identifiable {
    case moto, piedi
    var id: String { rawValue }
    var label: String { self == .piedi ? "A piedi" : "In moto" }
    var icon: String { self == .piedi ? "figure.walk" : "location.north.fill" }
    /// Velocità media stimata (km/h) per tracciati senza tempi (es. GPX)
    var avgSpeedKmh: Double { self == .piedi ? 5 : 45 }
}

struct MotoRoute: Codable, Identifiable, Hashable {
    var id: String
    var nome: String
    var partenza: String
    var arrivo: String
    var regione: String
    var km: Int
    var durataMin: Int
    var difficolta: String
    var stelle: Double
    var descrizione: String
    var tappe: [String]
    var waypointsGmaps: [String]
    var tags: [String]
    var fonte: String
    var stagione: String
    var isCustom: Bool?
    var legKm: [Int]?    // km per ogni tratta (disponibile solo per tragitti importati calcolati)
    var legMin: [Int]?   // minuti per ogni tratta
    var mezzo: String? = nil   // "moto" | "piedi" (nil = moto, per compatibilità)

    enum CodingKeys: String, CodingKey {
        case id, nome, partenza, arrivo, regione, km
        case durataMin, difficolta, stelle, descrizione
        case tappe, waypointsGmaps, tags, fonte, stagione, isCustom
        case legKm, legMin, mezzo
    }

    /// Mezzo del tragitto (default moto se non specificato)
    var transportMode: TransportMode { TransportMode(rawValue: mezzo ?? "moto") ?? .moto }

    var durataFormattata: String {
        let h = durataMin / 60
        let m = durataMin % 60
        if h == 0 { return "\(m) min" }
        if m == 0 { return "\(h)h" }
        return "\(h)h \(m)min"
    }

    var difficoltaColor: String {
        switch difficolta.lowercased() {
        case "facile":    return "green"
        case "difficile": return "red"
        default:          return "orange"
        }
    }

    func hash(into hasher: inout Hasher) { hasher.combine(id) }
    static func == (lhs: MotoRoute, rhs: MotoRoute) -> Bool { lhs.id == rhs.id }
}

struct RoutesResponse: Codable {
    var ok: Bool
    var routes: [MotoRoute]
    var total: Int?
    var error: String?
}

struct RegionsResponse: Codable {
    var ok: Bool
    var regions: [String]
    var error: String?
}

struct ComposeResponse: Codable {
    var ok: Bool
    var route: MotoRoute?
    var error: String?
}

struct ComposeRequest: Codable {
    var partenza: String?
    var arrivo: String?
    var regione: String?
    var kmMax: Int?
    var durataMax: Int?

    enum CodingKeys: String, CodingKey {
        case partenza, arrivo, regione
        case kmMax = "km_max"
        case durataMax = "durata_max"
    }
}

File: android/app/src/main/java/com/divito/inmoto/model/MotoRoute.kt
package com.divito.inmoto.model

import kotlinx.serialization.Serializable

/**
 * Mezzo con cui si percorre il tragitto: determina il tipo di routing (OSRM
 * driving/foot) e la taratura dei tempi di percorrenza. Port di TransportMode iOS.
 */
enum class TransportMode(val id: String, val label: String, val avgSpeedKmh: Double) {
    moto("moto", "In moto", 45.0),
    piedi("piedi", "A piedi", 5.0);

    companion object {
        fun from(value: String?): TransportMode =
            entries.firstOrNull { it.id == value } ?: moto
    }
}

/** Itinerario/tragitto moto. Mappa 1:1 il MotoRoute di iOS (stessi nomi JSON). */
@Serializable
data class MotoRoute(
    val id: String,
    val nome: String,
    val partenza: String,
    val arrivo: String,
    val regione: String,
    val km: Int,
    val durataMin: Int,
    val difficolta: String,
    val stelle: Double,
    val descrizione: String,
    val tappe: List<String> = emptyList(),
    val waypointsGmaps: List<String> = emptyList(),
    val tags: List<String> = emptyList(),
    val fonte: String = "",
    val stagione: String = "",
    val isCustom: Boolean? = null,
    val legKm: List<Int>? = null,   // km per ogni tratta (solo tragitti calcolati)
    val legMin: List<Int>? = null,  // minuti per ogni tratta
    val mezzo: String? = null,      // "moto" | "piedi" (null = moto, compatibilità)
) {
    /** Mezzo del tragitto (default moto se non specificato). */
    val transportMode: TransportMode get() = TransportMode.from(mezzo)

    val durataFormattata: String
        get() {
            val h = durataMin / 60
            val m = durataMin % 60
            return when {
                h == 0 -> "$m min"
                m == 0 -> "${h}h"
                else -> "${h}h ${m}min"
            }
        }

    /** "green" / "orange" / "red" — come iOS, mappato a colore nel layer UI. */
    val difficoltaColor: String
        get() = when (difficolta.lowercase()) {
            "facile" -> "green"
            "difficile" -> "red"
            else -> "orange"
        }
}

@Serializable
data class RoutesResponse(
    val ok: Boolean = false,
    val routes: List<MotoRoute> = emptyList(),
    val total: Int? = null,
    val error: String? = null,
)
